TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,2695 @@
+            if (throughProcedure != null)
+            {
+                Symbol throughProcedureSymbol = CheckSectionOrParagraph(p, throughProcedure);
+                if (throughProcedureSymbol == null)
+                    return false;
+                if (throughProcedureSymbol != procedureSymbol)
+                {
+                    if (procedureSymbol.Number > throughProcedureSymbol.Number)
+                    {// the second procedure name is before the first one.
+                        Diagnostic d = new Diagnostic(MessageCode.SemanticTCErrorInParser,
+                            p.CodeElement.Column,
+                            p.CodeElement.Column,
+                            p.CodeElement.Line,
+                            string.Format(Resource.BadPerformProcedureThru, procedure.ToString(), throughProcedure.ToString()));
+                        Diagnostics.Add(d);
+                        return false;
+                    }
+                    StoreProcedureSentenceBlocks(p, procedureSymbol, group, clonedBlockIndexMap);
+                    //Store all sentences or paragraphs between.
+                    for (int i = procedureSymbol.Number + 1; i < throughProcedureSymbol.Number; i++)
+                    {
+                        Symbol subSectionOrParagraph = this.CurrentProgramCfgBuilder.AllSectionsParagraphs[i];
+                        StoreProcedureSentenceBlocks(p, subSectionOrParagraph, group, clonedBlockIndexMap);
+                    }
+                    StoreProcedureSentenceBlocks(p, throughProcedureSymbol, group, clonedBlockIndexMap);
+                }
+                else
+                {
+                    StoreProcedureSentenceBlocks(p, procedureSymbol, group, clonedBlockIndexMap);
+                }
+            }
+            else
+            {
+                StoreProcedureSentenceBlocks(p, procedureSymbol, group, clonedBlockIndexMap);
+            }
+            //Now Clone the Graph.
+            if (!RelocateBasicBlockForNodeGroupGraph(p, group, clonedBlockIndexMap))
+            {
+                return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Relocate the graph generated by a BasicBlockForNodeGroup
+        /// </summary>
+        /// <param name="p">The Perform Procedure node source of the call</param>
+        /// <param name="group">The Group to be relocated</param>
+        /// <param name="clonedBlockIndexMap">The map between the old block indices to their new block indices.</param>
+        /// <returns>true if the relocation is successful, false if the relocation goes beyond the group limit, 
+        /// this often means that de target paragraphs of the PERFORM goes out of the paragraph.</returns>
+        private bool RelocateBasicBlockForNodeGroupGraph(PerformProcedure p, BasicBlockForNodeGroup group, Dictionary<int, int> clonedBlockIndexMap)
```

Comment:
I need to review this again. I don't get the meaning of the Diagnostic. Also we are talking about clonedBlock but it is an existing instance not a new one.

Reply:
Yes, at the time I though that blocks shall be cloned, but here the word "clone" shall be changed to "referenced". About diagnostics we shall use an option to emit or not diagnostics inside CFG, here it is a real CFG diagnostic

Reply:
Removed the 'clone' wording.